### PR TITLE
Some more SQL cleanup

### PR DESCRIPTION
**Pull request by erikrozendaal**
Reduce duplication by having a `latest_version` view. We may want to replace this with an actual table that we update whenever a new version becomes "active".

Replaces `NOT IN` conditions with `NOT EXISTS` (see https://wiki.postgresql.org/wiki/Don%27t_Do_This#Why_not.3F_5).

Simplified `DELETE` statements in `delete_old_versions`.

None of these changes should change the semantics.

Also, we can rewrite `delete_old_versions` to use a temporary table instead of a big common table expression:

```plpgsql
CREATE OR REPLACE FUNCTION delete_old_versions()
    RETURNS SETOF versions AS
$$
BEGIN
    CREATE TEMPORARY TABLE deleted_versions ON COMMIT DROP AS
        SELECT * FROM versions
         WHERE NOT EXISTS (SELECT * FROM reasonable_deltas d WHERE d.id = versions.id)
           AND NOT EXISTS (SELECT * FROM latest_version lv WHERE lv.id = versions.id);

    DELETE FROM versions WHERE id IN (SELECT id FROM deleted_versions);

    DELETE FROM object_log
     WHERE id <= (SELECT MAX(last_log_entry_id) FROM deleted_versions);

    DELETE FROM objects
     WHERE deleted_at IS NOT NULL
       AND NOT EXISTS (SELECT * FROM object_log WHERE new_object_id = objects.id)
       AND NOT EXISTS (SELECT * FROM object_log WHERE old_object_id = objects.id);

    -- Does not return from the function, but appends results to the return value:
    -- https://www.postgresql.org/docs/12/plpgsql-control-structures.html#id-1.8.8.8.3.4
    RETURN QUERY SELECT * FROM deleted_versions;

    DROP TABLE deleted_versions;
END
$$ LANGUAGE plpgsql;
```

### CODE REVIEW

File: src/main/resources/db/migration/R__functions.sql [hunk 1]
```diff
@@ -267,21 +267,21 @@ $body$
...
-WITH latest_version AS (
-    SELECT *
-    FROM versions
-    ORDER BY id DESC
-    LIMIT 1
-)
 SELECT z.*
 FROM (SELECT v.*,
-             SUM(delta_size) OVER (ORDER BY id DESC) AS total_delta_size
+             SUM(delta_size) OVER (PARTITION BY session_id ORDER BY id DESC) AS total_delta_size
```
Comment (lolepezy) [inline src/main/resources/db/migration/R__functions.sql:284]: Oh, that's a good catch!

**Review by lolepezy — APPROVED** [review]

**Comment by lolepezy** [comment]
> Also, we can rewrite delete_old_versions to use a temporary table instead of a big common table expression:

We can, no problems with that. 
Does it make things shorter/faster/look simpler?

I can imagine that the plan would be pretty much the same.

File: src/main/resources/db/migration/R__functions.sql (hunk 1 above)
Comment (lolepezy) [reply src/main/resources/db/migration/R__functions.sql:284]: Ah wait, there's `WHERE z.session_id = latest_version.session_id` so there's only one session id.

**Review by bjpbakker — APPROVED** [review]

**Comment by bjpbakker** [comment]
> Also, we can rewrite delete_old_versions to use a temporary table instead of a big common table expression

I'm not quite sure this would help (for either simplification or performance) but you can definitely try this (with a production-like workload).

**Review by lolepezy — APPROVED** [review]

**Pull request merged**